### PR TITLE
Flattening instances

### LINKED ISSUE #237: Support module instances in the simulator
This is an advanced feature. Not planned for MVP.

### LINKED ISSUE #528: Implement `instance` for constant expressions
It looks like `instance` in general may be too complex to implement #237. However, substituting constants with constant expressions is quite useful, e.g., for testing general protocol specs. We should prioritize this feature, while postponing #237.

It is not clear, at which stage this should be done. It's probably easy to do that at the resolver stage. However, this substitution would require the effects checker.

### PR DESCRIPTION
**Pull request by bugarela**
Hello :octocat: 

This finally closes #528 and we now have support for instances!
Also closes #237 (the sub-issue)

For now, we only implement flattening for instances. We should also flatten modules in the future in order to export a single flattened module to apalache.

- [x] Tests added for any new code
- [x] Ran `npm run format` (or had formatting run automatically on all files edited)
- [x] Documentation added for any new functionality
- [x] Entries added to the respective `CHANGELOG.md` for any new functionality
- [x] Feature table on [`README.md`](../README.md#roadmap) updated for any listed functionality

### CODE REVIEW

File: examples/cosmos/tendermint/TendermintAcc005.qnt [hunk 1]
```diff
@@ -55,32 +55,25 @@ module TendermintAcc {
...
-  pure val InvalidValues = Set("v2")
+  const InvalidValues: Set[Value_t]
 
   // the maximal round number
-  // const MaxRound: Round_t
-  pure val MaxRound = 4
+  const MaxRound: Round_t
 
   // the proposer function from 0..NRounds to the set of processes
   //const Proposer: Round_t -> Proc_t
```
[the remark this answers is not in the document]
Comment (bugarela) [reply examples/cosmos/tendermint/TendermintAcc005.qnt:79]: Right, I missed it

File: quint/src/runtime/compile.ts [hunk 2]
```diff
@@ -117,22 +120,35 @@ export function contextNameLookup(
...
-    const reorderedModules =
-      modules.filter(m => m.name !== mainName).concat(main ? [main] : [])
+    // Push back the main module to the end:
+    // The compiler exposes the state variables of the last module only.
+    const reorderedModules = flattenedModules.filter(m => m.name !== mainName).concat(main ? [main] : [])
     // Compile all modules
     reorderedModules.forEach(module => {
+      if (module.defs.some(d => d.kind === 'const')) {
+        // Skip modules with constants, as they are not used in the simulator
+        // They should be instanced in order to be evaluated
```
[the remark this answers is not in the document]
Comment (bugarela) [reply quint/src/runtime/compile.ts:149]: Oh my, will I ever get this word right? Sorry

File: quint/src/flattening.ts [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    case 'assume':
+      return {
+        ...def,
+        name: `${name}::${def.name}`,
+        assumption: addNamespaceToExpr(ctx, name, def.assumption),
+        id: ctx.idGenerator.nextId(),
+      }
+    case 'const':
+    case 'var':
+      return { ...def, name: `${name}::${def.name}`, id: ctx.idGenerator.nextId() }
```
[the remark this answers is not in the document]
Comment (bugarela) [reply quint/src/flattening.ts:108]: Yes, we have to something about the source maps, and this should be the best solution. I was planning that for a follow-up thing but forgot to open the issue. Will do that.

File: quint/src/flattening.ts [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+ *
+ * @param module The module to flatten
+ * @param table The lookup table to for all refered names
+ * @param modules The map of all refered modules
+ *
+ * @returns The flattened module
+ */
+export function flatten(module: QuintModule, table: LookupTable, modules: Map<string, QuintModule>): QuintModule {
+  const lastId = [...modules.values()].map(m => m.id).sort((a, b) => Number(a - b))[-1]
+  const idGenerator = newIdGenerator(lastId)
```
[the remark this answers is not in the document]
Comment (konnov) [reply quint/src/flattening.ts]: That's true. As far as I remember, `compile` and `compileAndTest` are receiving the generator already.

**Comment by bugarela** [comment]
> One thing I could not see is whether it works for transitive instances

@konnov It should work if they are flattened in the right order, as you said. If they are not, we throw this error:
https://github.com/informalsystems/quint/blob/4982bd23b1350c2d73f8796d689f355b9ec3a429/quint/src/flattening.ts#L120

We should test this and ensure we cover all scenarios. I'll register an issue.

**Pull request merged**